firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteIndexManager.java

```diff
@@ -88,21 +95,36 @@ public void start() {
       return;
     }
 
+    Map<Integer, FieldIndex.IndexState> indexStates = new HashMap<>();
+
+    // Fetch all index states if persisted for the user.
     db.query(
-            "SELECT index_id, collection_group, index_proto, update_time_seconds, "
-                + "update_time_nanos FROM index_configuration")
+            "SELECT index_id, sequence_number, read_time_seconds, read_time_nanos "
+                + "FROM index_state WHERE uid = ?")
+        .forEach(
+            row -> {
+              int indexId = row.getInt(1);
+              long sequenceNumber = row.getLong(1);
+              SnapshotVersion readTime =
+                  new SnapshotVersion(new Timestamp(row.getLong(2), row.getInt(3)));
+              indexStates.put(indexId, FieldIndex.IndexState.create(sequenceNumber, readTime));
+            });
+
+    // Fetch all indices and combine with user's index state if available.
+    db.query("SELECT index_id, collection_group, index_proto FROM index_configuration")
         .forEach(
             row -> {
               try {
                 int indexId = row.getInt(0);
                 String collectionGroup = row.getString(1);
                 List<FieldIndex.Segment> segments =
                     serializer.decodeFieldIndexSegments(Index.parseFrom(row.getBlob(2)));
-                SnapshotVersion updateTime =
-                    new SnapshotVersion(new Timestamp(row.getLong(3), row.getInt(4)));
-
+                FieldIndex.IndexState indexState =
```

Comment:
optional: can you add a comment here that describes why we need this?

Initially I thought it was strange that we didn't just set the index state directly when we created a field index, and instead had this misdirection and check. After thinking about it, i realized that a new user might have not have an index state created for the existing field index, which is why we need this, but I'm afraid I'll forget this by next week.